default/python/universe_generation/starsystems.py

```diff
@@ -70,6 +70,16 @@ def name_planets(system):
         fo.set_name(planet, name)
 
 
+def can_have_no_planets(star_type):
+    """
+    Return True is system is allowed to have no planets.
+    This intentionally only checks the STAR_TYPE_MOD_TO_PLANET_SIZE_DIST so that
+    the following code forces most stars to have a planetary body so that in the
+    GUI the presence of a star implies a planet
+    """
+    return universe_tables.STAR_TYPE_MOD_TO_PLANET_SIZE_DIST[star_type][fo.planetSize.noWorld] != 0
```

Comment:
Hm, the original purpose of avoiding star systems (which are not deep space nodes) with no planets is to reduce annoying micromanagement (clicking through a lot of star systems to find the ones actually containing planets). Making an exception for black hole systems is reasonable, because they are sufficiently rare. Neutron stars, while still not as common as the other star types, are certainly more common than black holes.

But probably still uncommon enough so that probably won't be too much of a problem. That, and with the suggested comments in `universe_tables.py`, the fix for checking `> 0` I guess we can leave it at that. If things doesn't work well enough, we can make further adjustments later.

> An "engine" should deal with the "source material", in this case universe_tables.py honestly.

What exactly do you mean by that?